File: sdks/rust_sdk/reference-services/src/agents/run_policy.rs
//! Shared run-command policy for the reference agents (campaign R2).
//!
//! Replaces shell-string execution of message-derived
//! commands with an argv allowlist, strict validation, fail-closed LLM
//! confirmation, and an explicit operator-confirmation flag. Shell compounds
//! are rejected, never re-interpreted.

use serde_json::Value;

pub const ALLOWED_INTERPRETERS: [&str; 5] = ["python3", "python", "node", "npm", "npx"];
const MAX_ARGV_LENGTH: usize = 16;
const MAX_ARG_LENGTH: usize = 512;
const FORBIDDEN: &[char] = &[
    ';', '&', '|', '`', '$', '<', '>', '(', ')', '{', '}', '*', '"', '\'', '\n', '\r', '\\',
];

fn arg_is_safe(arg: &str) -> Result<(), String> {
    if arg.is_empty() {
        return Err("run command arguments must be non-empty strings".into());
    }
    if arg.len() > MAX_ARG_LENGTH {
        return Err(format!(
            "run command argument too long: {}…",
            &arg[..32.min(arg.len())]
        ));
    }
    if arg.contains(FORBIDDEN) {
        return Err(format!("run command argument contains forbidden characters: {arg:?}"));
    }
    if arg.starts_with('/') {
        return Err(format!("absolute paths are not allowed in run commands: {arg:?}"));
    }
    if arg == ".." || arg.starts_with("../") || arg.contains("/../") || arg.ends_with("/..") {
        return Err(format!("path traversal is not allowed in run commands: {arg:?}"));
    }
    Ok(())
}

/// Validate a full argv. Returns the validated argv or an explanation.
pub fn validate_run_argv(argv: &[String]) -> Result<Vec<String>, String> {
    if argv.is_empty() {
        return Err("run command must be a non-empty argv array".into());
    }
    if argv.len() > MAX_ARGV_LENGTH {
        return Err(format!("run command exceeds {MAX_ARGV_LENGTH} arguments"));
    }
    for arg in argv {
        arg_is_safe(arg)?;
    }
    let interpreter = &argv[0];
    if !ALLOWED_INTERPRETERS.contains(&interpreter.as_str()) {
        return Err(format!(
            "interpreter {interpreter:?} is not in the allowlist ({})",
            ALLOWED_INTERPRETERS.join(", ")
        ));
    }
    Ok(argv.to_vec())
}

/// Parse a legacy single-string command strictly. Shell compounds are
/// rejected (never re-interpreted).
pub fn parse_run_command(raw: &str) -> Result<Vec<String>, String> {
    let cmd = raw.trim();
    if cmd.is_empty() {
        return Err("run command is empty".into());
    }
    if cmd.contains(FORBIDDEN) {
        return Err("run command contains forbidden characters (shell compounds are rejected)".into());
    }
    let argv: Vec<String> = cmd.split_whitespace().map(String::from).collect();
    validate_run_argv(&argv)
}

pub fn validate_run_cmd_string(raw: &str) -> Result<Vec<String>, String> {
    parse_run_command(raw)
}

/// Resolve the confirmed argv for a run stage from `params`.
///
/// - `params.cmd_argv` (validated argv) is used as the operator-provided path.
/// - otherwise `params.cmd` is LLM-confirmed; any confirmation failure is
///   terminal (the raw suggestion is never executed).
/// - `params.confirm` must be exactly `true`.
///
/// Returns `Ok((argv, display))` or `Err(error_info_json)`.
pub fn resolve_run_stage(params: &Value) -> Result<(Vec<String>, Value), Value> {
    let confirm = params.get("confirm").and_then(|v| v.as_bool()).unwrap_or(false);
    if !confirm {
        return Err(serde_json::json!({"error": "run_not_confirmed"}));
    }
    if let Some(argv_value) = params.get("cmd_argv") {
        let parsed: Vec<String> = match serde_json::from_value(argv_value.clone()) {
            Ok(v) => v,
            Err(_) => {
                return Err(serde_json::json!({"error": "run command must be an array of strings"}));
            }
        };
        return match validate_run_argv(&parsed) {
            Ok(argv) => Ok((argv.clone(), serde_json::json!({"cmd": argv.join(" "), "argv": argv}))),
            Err(e) => Err(serde_json::json!({"error": e})),
        };
    }
    let suggested = params.get("cmd").and_then(|v| v.as_str()).unwrap_or("");
    if suggested.trim().is_empty() {
        return Err(serde_json::json!({"error": "no_run_command"}));
    }
    // LLM confirmation happens in the caller (async runtime); this function
    // validates the confirmed string.
    match validate_run_cmd_string(suggested) {
        Ok(argv) => {
            let display = serde_json::json!({"cmd": suggested, "argv": argv});
            Ok((argv, display))
        }
        Err(e) => Err(serde_json::json!({"error": e, "cmd": suggested})),
    }
}

pub fn confirmation_failed(cmd: &str) -> Value {
    serde_json::json!({"error": "llm_confirmation_unavailable", "cmd": cmd})
}

File: sdks/js_sdk/reference-services/agents/run_policy.ts
// Shared run-command policy for the reference agents (campaign R2).
//
// The run stage used to execute message-derived strings through a shell
// spawn with an unmediated fallback to the raw suggestion. This
// module replaces that with:
//   - an argv allowlist (no shell anywhere),
//   - strict validation of every argument,
//   - fail-closed behavior when LLM confirmation is unavailable or its output
//     is unusable (the raw suggestion is never executed),
//   - an explicit operator-confirmation flag required before any spawn.
//
// Legacy single-string commands are accepted only when they survive a strict
// tokenizer plus the same allowlist — shell compounds like
// "cd backend && python3 server.py" are rejected, never re-interpreted.

export const ALLOWED_INTERPRETERS: ReadonlySet<string> = new Set([
  'python3',
  'python',
  'node',
  'npm',
  'npx',
]);

// Characters that make an argument shell- or path-hostile. The argv form is
// executed without a shell, so these are rejected defensively (and make
// legacy string parsing fail closed rather than approximate).
const FORBIDDEN = /[;&|`$<>(){}*"'\n\r\\]/;
const MAX_ARGV_LENGTH = 16;
const MAX_ARG_LENGTH = 512;

export type RunArgv = { ok: true; argv: string[] };
export type RunRejection = { ok: false; error: string };
export type RunResolution = RunArgv | RunRejection;

export function validateRunArgv(argv: unknown): RunResolution {
  if (!Array.isArray(argv) || argv.length === 0) {
    return { ok: false, error: 'run command must be a non-empty argv array' };
  }
  if (argv.length > MAX_ARGV_LENGTH) {
    return { ok: false, error: `run command exceeds ${MAX_ARGV_LENGTH} arguments` };
  }
  for (const raw of argv) {
    if (typeof raw !== 'string' || raw.length === 0) {
      return { ok: false, error: 'run command arguments must be non-empty strings' };
    }
    if (raw.length > MAX_ARG_LENGTH) {
      return { ok: false, error: `run command argument too long: ${raw.slice(0, 32)}…` };
    }
    if (FORBIDDEN.test(raw)) {
      return { ok: false, error: `run command argument contains forbidden characters: ${JSON.stringify(raw)}` };
    }
    if (raw.startsWith('/')) {
      return { ok: false, error: `absolute paths are not allowed in run commands: ${JSON.stringify(raw)}` };
    }
    if (raw === '..' || raw.startsWith('../') || raw.includes('/../') || raw.endsWith('/..')) {
      return { ok: false, error: `path traversal is not allowed in run commands: ${JSON.stringify(raw)}` };
    }
  }
  const interpreter = argv[0];
  if (!ALLOWED_INTERPRETERS.has(interpreter)) {
    return {
      ok: false,
      error: `interpreter ${JSON.stringify(interpreter)} is not in the allowlist (${[...ALLOWED_INTERPRETERS].join(', ')})`,
    };
  }
  return { ok: true, argv: argv as string[] };
}

export function parseRunCommand(raw: string): RunResolution {
  const cmd = typeof raw === 'string' ? raw.trim() : '';
  if (!cmd) return { ok: false, error: 'run command is empty' };
  if (FORBIDDEN.test(cmd)) {
    return { ok: false, error: 'run command contains forbidden characters (shell compounds are rejected)' };
  }
  return validateRunArgv(cmd.split(/\s+/).filter(part => part.length > 0));
}

export function validateRunCmdString(raw: string): RunResolution {
  return parseRunCommand(raw);
}

export interface RunStageParams {
  cmd?: unknown;
  cmd_argv?: unknown;
  confirm?: unknown;
}

export interface RunStageDeps {
  /** LLM confirmation callback (mockable in tests). */
  runClaude: (prompt: string) => Promise<any>;
  /** Process spawner (mockable in tests); MUST NOT be given a shell option. */
  spawnFn: (
    program: string,
    args: string[],
    options: { cwd: string; stdio: ['ignore', unknown, unknown] },
  ) => { pid?: number };
  baseDir: string;
  logPath: string;
  writePid: (pid: string) => void;
  appendLog: () => unknown;
}

export interface RunStageResult {
  status: 'ok' | 'error';
  info: Record<string, unknown>;
}

/**
 * Execute the confirmed run stage, or fail closed. Never spawns a shell and
 * never falls back to the raw suggestion when LLM confirmation is missing.
 */
export async function executeRunStage(
  params: RunStageParams,
  deps: RunStageDeps,
): Promise<RunStageResult> {
  if (params.confirm !== true) {
    return { status: 'error', info: { error: 'run_not_confirmed' } };
  }

  let resolution: RunResolution;
  if (params.cmd_argv !== undefined) {
    resolution = validateRunArgv(params.cmd_argv);
    if (!resolution.ok) return { status: 'error', info: { error: resolution.error } };
  } else if (typeof params.cmd === 'string' && params.cmd.trim()) {
    // Confirm-then-run: the LLM reviews the suggestion and returns the argv.
    // Any failure is terminal — the raw suggestion is never executed.
    const suggested = params.cmd;
    let resultObj: any;
    try {
      resultObj = await deps.runClaude(buildConfirmationPrompt(suggested));
    } catch {
      return { status: 'error', info: { error: 'llm_confirmation_unavailable', cmd: suggested } };
    }
    const runCmd = resultObj && typeof resultObj.run_cmd === 'string' ? resultObj.run_cmd : undefined;
    if (!runCmd) {
      return { status: 'error', info: { error: 'llm_confirmation_unavailable', cmd: suggested } };
    }
    resolution = validateRunCmdString(runCmd);
    if (!resolution.ok) return { status: 'error', info: { error: resolution.error, cmd: runCmd } };
  } else {
    return { status: 'error', info: { error: 'no_run_command' } };
  }

  const argv = (resolution as RunArgv).argv;
  try {
    const out = deps.appendLog();
    const child = deps.spawnFn(argv[0], argv.slice(1), {
      cwd: deps.baseDir,
      stdio: ['ignore', out, out] as ['ignore', unknown, unknown],
    });
    deps.writePid(String(child.pid ?? ''));
    return { status: 'ok', info: { cmd: argv.join(' '), argv, pid: child.pid } };
  } catch (e: any) {
    return { status: 'error', info: { error: String(e?.message || e), argv } };
  }
}

export function buildConfirmationPrompt(suggested: string): string {
  return (
    'You are the run orchestrator. A suggested command was provided. ' +
    'Return JSON ONLY as {"run_cmd": string}. Execution cwd is ./generated_app. ' +
    'The command must be a single program with arguments (no shell compounds, no cd). ' +
    `Suggested: ${JSON.stringify(suggested)}`
  );
}
